extensions/azuremonitor/src/extension.ts

```diff
@@ -0,0 +1,137 @@
+import { AzureMonitorServer } from './azuremonitorServer';
+import { promises as fs } from 'fs';
+
+const localize = nls.loadMessageBundle();
+
+export async function activate(context: vscode.ExtensionContext): Promise<IExtension | undefined> {
+	// lets make sure we support this platform first
+	let supported = await Utils.verifyPlatform();
+
+	if (!supported) {
+		vscode.window.showErrorMessage(localize('azuremonitor.unsupportedPlatform', 'Unsupported platform'));
+		return undefined;
+	}
+
+	// ensure our log path exists
+	if (!(await Utils.exists(context.logPath))) {
+		await fs.mkdir(context.logPath);
+	}
+
+	let appContext = new AppContext(context);
+
+	let nodeProvider = new AzureMonitorObjectExplorerNodeProvider(appContext);
+	azdata.dataprotocol.registerObjectExplorerNodeProvider(nodeProvider);
+	let iconProvider = new AzureMonitorIconProvider();
+	azdata.dataprotocol.registerIconProvider(iconProvider);
+
+	activateNotebookTask();
+
+	registerSearchServerCommand();
+	context.subscriptions.push(new ContextProvider());
+
+	registerLogCommand(context);
+
+	// initialize client last so we don't have features stuck behind it
+	const server = new AzureMonitorServer();
+	context.subscriptions.push(server);
+	await server.start(appContext);
+
+	return createAzureMonitorApi(appContext);
+}
+
+const logFiles = ['resourceprovider.log', 'azuremonitorservice.log', 'credentialstore.log'];
+function registerLogCommand(context: vscode.ExtensionContext) {
+	context.subscriptions.push(vscode.commands.registerCommand('azuremonitor.showLogFile', async () => {
+		const choice = await vscode.window.showQuickPick(logFiles);
+		if (choice) {
+			const document = await vscode.workspace.openTextDocument(vscode.Uri.file(path.join(context.logPath, choice)));
+			if (document) {
+				vscode.window.showTextDocument(document);
+			}
+		}
+	}));
+}
+
+function activateNotebookTask(): void {
+	azdata.tasks.registerTask(Constants.azuremonitorClusterNewNotebookTask, (profile: azdata.IConnectionProfile) => {
+		return saveProfileAndCreateNotebook(profile);
+	});
+	azdata.tasks.registerTask(Constants.azuremonitorClusterOpenNotebookTask, (profile: azdata.IConnectionProfile) => {
+		return handleOpenNotebookTask(profile);
+	});
+}
+
+function saveProfileAndCreateNotebook(profile: azdata.IConnectionProfile): Promise<void> {
+	return handleNewNotebookTask(undefined, profile);
+}
+
+function findNextUntitledEditorName(): string {
+	let nextVal = 0;
+	// Note: this will go forever if it's coded wrong, or you have inifinite Untitled notebooks!
+	while (true) {
+		let title = `Notebook-${nextVal}`;
+		let hasNotebookDoc = azdata.nb.notebookDocuments.findIndex(doc => doc.isUntitled && doc.fileName === title) > -1;
+		if (!hasNotebookDoc) {
+			return title;
+		}
+		nextVal++;
+	}
+}
+
+async function handleNewNotebookTask(_oeContext?: azdata.ObjectExplorerContext, profile?: azdata.IConnectionProfile): Promise<void> {
```

Comment:
@chlafreniere Do you know of any built-in functionality to do this stuff? Seems annoying to have to have extensions implement this generic kind of stuff themselves every time.